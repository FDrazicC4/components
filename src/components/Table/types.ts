import { ReactNode } from 'react';
import { Props as ColumnProps } from './components/TableHeader/components/TableHeaderColumn/types';
import { RowData } from './components/TableRow/types';


export interface Props {


    /**
     * The custom JSX element to use for the Table (otherwise, it uses `<table />`)
     */
    as?: keyof JSX.IntrinsicElements | null,


    /**
     * Any custom content to fill the TableHeader that is different than default
     */
    children?: ReactNode,


    /**
     * The className for the TableHeader
     */
    className?: string,


    /**
     * The column data used to populate the Table's header
     */
    columnConfig?: ColumnProps[],


    /**
     * The row data used to populate the TableBody
     */
    rows?: RowData[],


}
