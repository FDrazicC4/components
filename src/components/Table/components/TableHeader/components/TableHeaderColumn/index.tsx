import React, { ReactElement } from 'react';
import clsx from 'clsx';
import { Props } from './types';


/**
 * The `<TableHeaderColumn />` component represents a single column in the `<TableHeader />`.
 */
function TableHeaderColumn({
    children,
    className,
    label,
    ...otherProps
}: Props): ReactElement {


    const classes = clsx(
        'brew-TableHeaderColumn',
        className,
    );


    const content = children || label;


    return (
        <th className={classes} {...otherProps}>
            {content}
        </th>
    );


}


export default TableHeaderColumn;
