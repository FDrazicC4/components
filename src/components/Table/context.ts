import * as React from 'react';

export const TableLevel2Context = React.createContext<{ variant: 'head' | 'body'} | undefined>(undefined);