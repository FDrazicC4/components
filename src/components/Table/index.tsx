import React from 'react';
import clsx from 'clsx';
import { Props } from './types';
import MuiTable from '@material-ui/core/Table';


const Table = React.forwardRef(({ className, ...props }: Props, ref: React.Ref<any>): React.ReactElement => {


    const classes = clsx(
        'brew-Table',
        className,
    );

    return (
        <MuiTable className={classes} ref={ref} {...props} />
    );

});


Table.displayName = 'BrewkitTable';

export default Table;
