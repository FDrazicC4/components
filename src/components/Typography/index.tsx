import React from 'react';
import clsx from 'clsx';
import { Typography as MuiTypography } from '@material-ui/core';
import { Props } from './types';



/**
 * The `<Typography />` component renders text using the 'Roboto' font in various styles.
 */
export const Typography = React.forwardRef(({
    variant = 'body1',
    align = 'inherit',
    display = 'initial',
    shouldTruncate = false,
    color,
    hasGutter = false,
    className,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const classNames = clsx(
        'brew-Typography',
        [`brew-Typography--display-${display}`],
        { [`brew-Typography--variant-${String(variant)}`]: variant },
        { [`brew-Typography--align-${String(align)}`]: align },
        { [`brew-Typography--color-${String(color)}`]: color },
        { 'brew-Typography--shouldTruncate': shouldTruncate },
        { 'brew-Typography--hasGutter': hasGutter },
        className,
    );


    return (
        <MuiTypography
            align={align}
            className={classNames}
            display={display}
            gutterBottom={hasGutter}
            noWrap={shouldTruncate}
            ref={ref}
            variant={variant}
            {...otherProps}
        />
    );


});


Typography.displayName = 'Typography';


export default Typography;
