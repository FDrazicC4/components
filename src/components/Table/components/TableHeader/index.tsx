import React, { ReactElement } from 'react';
import clsx from 'clsx';
import TableHeaderColumn from './components/TableHeaderColumn';
import { Props as ColumnProps } from './components/TableHeaderColumn/types';
import TableRow from '../TableRow';
import { Props } from './types';


/**
 * The `<TableHeader />` component represents the standard header of a `<Table />` component.
 */
function TableHeader({
    children,
    className,
    columns,
    ...otherProps
}: Props): ReactElement {


    const classes = clsx(
        'brew-TableHeader',
        className,
    );


    const content = children || (columns?.map((colData: ColumnProps, index: number) => (
        <TableHeaderColumn key={index} {...colData} />
    )));


    return (
        <thead className={classes} {...otherProps}>
            <TableRow>
                {content}
            </TableRow>
        </thead>
    );


}

TableHeader.Column = TableHeaderColumn;

export default TableHeader;
