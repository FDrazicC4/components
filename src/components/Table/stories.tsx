import React, { ReactElement, ReactNode } from 'react';
import { object, text } from '@storybook/addon-knobs';
import Label from '../Label';
import { Props as ColumnProps } from './components/TableHeader/components/TableHeaderColumn/types';
import Table from './index';


export default {
    component: Table,
    parameters: {
        componentSubtitle: <Label color="warning">In Development</Label>,
    },
    title: 'Components|Display/Table',
};


export const General = (): ReactElement => {


    const colExample = [
        { label: 'Column 1' },
        { label: 'Column 2' },
        { label: 'Column 3' },
    ];

    const rowExample = [
        ['row1-col1', 'row1-col2', 'row1-col3'],
        ['row2-col1', 'row2-col2', 'row2-col3'],
        ['row3-col1', 'row3-col2', 'row3-col3'],
    ];

    const className: string = text('className', 'churros');
    const children: ReactNode = text('children', '');
    const columns: ColumnProps[] = object('columns', colExample);
    const rows: ReactNode[][] = object('rows', rowExample);


    return (
        <Table
            className={className}
            columns={columns}
            rows={rows}
        >
            {children}
        </Table>
    );
};
