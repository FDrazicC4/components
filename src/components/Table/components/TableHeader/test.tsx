import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Table from '../../index';
import TableHeader from './index';
import TableHeaderColumn from './components/TableHeaderColumn';
import Typography from '../../../Typography';


describe('Table Header', () => {


    it('Renders correctly', () => {
        const header = renderer
            .create((
                <TableHeader>
                    <TableHeaderColumn>Column1</TableHeaderColumn>
                    <TableHeaderColumn>Column2</TableHeaderColumn>
                    <TableHeaderColumn>Column3</TableHeaderColumn>
                </TableHeader>
            )).toJSON();

        expect(header).toMatchSnapshot();
    });


    it('Includes custom className', () => {
        const className = `class-${String(Math.floor(Math.random() * 1000))}`;
        const content = shallow((
            <Table>
                <TableHeader className={className}>
                    <TableHeaderColumn>
                        Column4
                    </TableHeaderColumn>
                    <TableHeaderColumn>
                        Column5
                    </TableHeaderColumn>
                    <TableHeaderColumn>
                        Column6
                    </TableHeaderColumn>
                </TableHeader>
            </Table>
        ));

        expect(content.children().first()
            .shallow()
            .hasClass(className))
            .toBe(true);
        expect(content.children().first()
            .shallow()
            .hasClass('brew-TableHeader'))
            .toBe(true);
    });


});

describe('Table Header Props', () => {


    it('Renders "columns" prop correctly', () => {
        const header = renderer
            .create((
                <TableHeader
                    columns={[
                        { label: 'ColumnA' },
                        { label: 'ColumnB' },
                        { label: 'ColumnC' },
                    ]}
                />
            ))
            .toJSON();

        expect(header).toMatchSnapshot();

        expect(mount((
            <Table>
                <TableHeader
                    columns={[
                        { label: <Typography variant="h3">ColumnA</Typography> },
                        { label: <Typography variant="h3">ColumnB</Typography> },
                        { label: <Typography variant="h3">ColumnC</Typography> },
                    ]}
                />
            </Table>
        ))
            .find(Typography)
            .find('[variant="h3"]')
            .children()
            .contains('ColumnB')).toBe(true);
    });

});
