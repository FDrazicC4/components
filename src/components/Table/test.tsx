import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Table from './index';
import TableBody from './components/TableBody';
import TableHeader from './components/TableHeader';
import TableHeaderColumn from './components/TableHeader/components/TableHeaderColumn';
import TableRow from './components/TableRow';
import Typography from '../Typography';


describe('Table', () => {

    it('Renders correctly', () => {
        const content = renderer
            .create((
                <Table>
                    <TableHeader>
                        <TableHeaderColumn>Column1</TableHeaderColumn>
                        <TableHeaderColumn>Column2</TableHeaderColumn>
                        <TableHeaderColumn>Column3</TableHeaderColumn>
                    </TableHeader>
                    <TableBody>
                        <TableRow>
                            <td>Row1-Col1</td><td>Row1-Col2</td><td>Row1-Col3</td>
                        </TableRow>
                        <TableRow>
                            <td>Row2-Col1</td><td>Row2-Col2</td><td>Row2-Col3</td>
                        </TableRow>
                    </TableBody>
                </Table>
            ))
            .toJSON();
        expect(content).toMatchSnapshot();
    });

    it('Includes custom className', () => {
        const className = `class-${String(Math.floor(Math.random() * 1000))}`;
        const content = shallow((
            <Table className={className}>
                <TableHeader>
                    <TableHeaderColumn>Column1</TableHeaderColumn>
                    <TableHeaderColumn>Column2</TableHeaderColumn>
                    <TableHeaderColumn>Column3</TableHeaderColumn>
                </TableHeader>
                <TableBody>
                    <TableRow>
                        <td>Row1-Col1</td><td>Row1-Col2</td><td>Row1-Col3</td>
                    </TableRow>
                    <TableRow>
                        <td>Row2-Col1</td><td>Row2-Col2</td><td>Row2-Col3</td>
                    </TableRow>
                </TableBody>
            </Table>
        ));

        expect(content.hasClass(className)).toBe(true);
        expect(content.hasClass('brew-Table')).toBe(true);
    });


});

// eslint-disable-next-line max-lines-per-function
describe('Table Props', () => {


    it('Renders "columns" prop correctly', () => {
        const content = renderer
            .create((
                <Table
                    columns={[
                        { label: 'ColumnA' },
                        { label: 'ColumnB' },
                        { label: 'ColumnC' },
                    ]}
                />
            ))
            .toJSON();
        expect(content).toMatchSnapshot();

        expect(mount((
            <Table
                columns={[
                    { label: <Typography variant="body1">ColumnA</Typography> },
                    { label: <Typography variant="body1">ColumnB</Typography> },
                    { label: <Typography variant="body1">ColumnC</Typography> },
                ]}
            />
        ))
            .find(Typography)
            .find('[variant="body1"]')
            .children()
            .contains('ColumnA')).toBe(true);
    });


    it('Renders "rows" prop correctly', () => {
        const content = renderer
            .create((
                <Table
                    rows={[
                        ['row1-col1', 'row1-col2', 'row1-col3'],
                        ['row2-col1', 'row2-col2', 'row2-col3'],
                        ['row3-col1', 'row3-col2', 'row3-col3'],
                    ]}
                />
            ))
            .toJSON();
        expect(content).toMatchSnapshot();

        expect(mount((
            <Table
                rows={[
                    [
                        <Typography key="row1-col1" variant="body2">Row1-Col1</Typography>,
                        <Typography key="row1-col2" variant="body2">Row1-Col2</Typography>,
                        <Typography key="row1-col3" variant="body2">Row1-Col3</Typography>,
                    ],
                    [
                        <Typography key="row2-col1" variant="body2">Row2-Col1</Typography>,
                        <Typography key="row2-col2" variant="body2">Row2-Col2</Typography>,
                        <Typography key="row2-col3" variant="body2">Row2-Col3</Typography>,
                    ],
                    [
                        <Typography key="row3-col1" variant="body2">Row3-Col1</Typography>,
                        <Typography key="row3-col2" variant="body2">Row3-Col2</Typography>,
                        <Typography key="row3-col3" variant="body2">Row3-Col3</Typography>,
                    ],
                ]}
            />
        ))
            .find(Typography)
            .find('[variant="body2"]')
            .children()
            .contains('Row3-Col3')).toBe(true);
    });

    it('Renders "columns" and "rows" prop correctly simultaneously', () => {
        const content = renderer
            .create((
                <Table
                    columns={[
                        { label: 'ColumnX' },
                        { label: 'ColumnY' },
                        { label: 'ColumnZ' },
                    ]}
                    rows={[
                        ['row1-colX', 'row1-colX', 'row1-colX'],
                        ['row2-colY', 'row2-colY', 'row2-colY'],
                        ['row3-colZ', 'row3-colZ', 'row3-colZ'],
                    ]}
                />
            ))
            .toJSON();
        expect(content).toMatchSnapshot();

        expect(mount((
            <Table
                columns={[
                    { label: <Typography variant="h3">ColumnX</Typography> },
                    { label: <Typography variant="h3">ColumnY</Typography> },
                    { label: <Typography variant="h3">ColumnZ</Typography> },
                ]}
                rows={[
                    [
                        <Typography key="row1-colx" variant="body2">Row1-ColX</Typography>,
                        <Typography key="row1-coly" variant="body2">Row1-ColY</Typography>,
                        <Typography key="row1-colz" variant="body2">Row1-ColZ</Typography>,
                    ],
                    [
                        <Typography key="row2-colx" variant="body2">Row2-ColX</Typography>,
                        <Typography key="row2-coly" variant="body2">Row2-ColY</Typography>,
                        <Typography key="row2-colz" variant="body2">Row2-ColZ</Typography>,
                    ],
                    [
                        <Typography key="row3-colx" variant="body2">Row3-ColX</Typography>,
                        <Typography key="row3-coly" variant="body2">Row3-ColY</Typography>,
                        <Typography key="row3-colz" variant="body2">Row3-ColZ</Typography>,
                    ],
                ]}
            />
        ))
            .find(Typography)
            .find('[variant="body2"]')
            .children()
            .contains('Row2-ColY')).toBe(true);
    });


});
