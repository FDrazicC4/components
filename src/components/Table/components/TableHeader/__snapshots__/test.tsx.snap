// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Header Props Renders "columns" prop correctly 1`] = `
<thead
  className="brew-TableHeader"
>
  <tr
    className="brew-TableRow"
  >
    <th
      className="brew-TableHeaderColumn"
    >
      ColumnA
    </th>
    <th
      className="brew-TableHeaderColumn"
    >
      ColumnB
    </th>
    <th
      className="brew-TableHeaderColumn"
    >
      ColumnC
    </th>
  </tr>
</thead>
`;

exports[`Table Header Renders correctly 1`] = `
<thead
  className="brew-TableHeader"
>
  <tr
    className="brew-TableRow"
  >
    <th
      className="brew-TableHeaderColumn"
    >
      Column1
    </th>
    <th
      className="brew-TableHeaderColumn"
    >
      Column2
    </th>
    <th
      className="brew-TableHeaderColumn"
    >
      Column3
    </th>
  </tr>
</thead>
`;
